### PR TITLE
cache plone.restapi and introduce a terse dynamic content short time cache

### PR DESCRIPTION
**Pull request by jensens**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -67,6 +67,7 @@
         'test': [
             'plone.app.contenttypes[test]',
             'plone.app.testing',
+            'plone.restapi[testing]',
```
Comment (mamico) [inline setup.py:70]: typo?
```
- 'plone.restapi[testing]',
+ 'plone.restapi[test]',
```

**Comment by mamico** [comment]
My two cents.

I have some doubt about the implementation, my test (using this branch and plone 5.2 with python 3):

* I created a plone site
* I have installed plone.restapi add-on
* activated p.a.caching
* imported the caching profile "with caching proxy"
* added an image

The first content request has a response like:

```
% wget -O /dev/null -S http://127.0.0.2:8080/Plone/2725159511543238876-128.png
  HTTP/1.1 200 OK
  Cache-Control: max-age=0, s-maxage=86400, must-revalidate
  Content-Type: image/png
  Date: Sun, 01 Nov 2020 23:48:35 GMT
  Expires: Thu, 04 Nov 2010 23:48:35 GMT
  Last-Modified: Sun, 01 Nov 2020 23:47:51 GMT
  X-Cache-Operation: plone.app.caching.moderateCaching
  X-Cache-Rule: plone.content.file
  ...
```

Now, because of s-maxage=86400, if I do a REST API request

`% wget -O /dev/null --header "Accept: application/json" -S http://127.0.0.2:8080/Plone/2725159511543238876-128.png
`
I haven't a JSON response as I expected but the image that was cached by the proxy, that is because there is no cache key in proxy other than URL.

I had not a better result with the profile "without caching proxy"

The first "normal" request, opening the URL inside a browser tab,  has a response like:

```
  HTTP/1.1 200 OK
  Cache-Control: max-age=0, must-revalidate, private
  Content-Type: image/png
  Date: Sun, 01 Nov 2020 23:48:35 GMT
  Expires: Thu, 04 Nov 2010 23:48:35 GMT
  Last-Modified: Sun, 01 Nov 2020 23:47:51 GMT
  X-Cache-Operation: plone.app.caching.weakCaching
  X-Cache-Rule: plone.content.file
```

A javascript request for restapi:

```
var req = new XMLHttpRequest();
req.open("GET", "http://127.0.0.2:8080/Plone/2725159511543238876-128.png")
req.setRequestHeader('Accept', 'application/json')
req.send()
```

produce a browser request like:

```
Accept: application/json
Accept-Encoding: gzip, deflate
If-Modified-Since: Sun, 01 Nov 2020 23:47:51 GMT
...
```

and "304 Not Modified" response by Plone (so the browser consider the cached png like a valid response)

My example above is about image content, but I think that the same problem could be also with other content basically because on the client cache side the key for the cache is only the URL (no Vary header) and the REST API request has the same URL than browser requests.

I don't know if I miss some steps, but I did some experiments a time ago and my conclusion was that possible alternatives were:

1. add "Vary: Accept" header to responses, with the risk of wasting proxy caching because for HTML request every browser has a different Accept header
2. ... or (IMHO could be the simpler/best alternative) add a custom header to all client rest API request like "X-ApiRequest: 1" and a "Vary: X-ApiRequest" 
3. ... or have a different URL for restapi than browser request
4. ... or do the job on proxy caching [1] and in p.a.caching for 304 responses logic.
5. ... or any other better idea


BTW: There is a possible bug using VAry header in the response and RamCache, I tried to address this here https://github.com/plone/plone.app.caching/tree/vary-ramcache
time ago, no time to complete/review that, if you agree with the problem and want to review, I could create a PR from that.

[1] Varnish rules for manage with option 4
```
sub vcl_recv {
  ...
 if (req.http.Accept ~ "application/json") {
    set req.http.X-RestApi = "1";
 }

sub vcl_fetch {
  ...
  if (beresp.http.Vary) {
    set beresp.http.Vary = beresp.http.Vary ", X-RestApi";
  } else {
    set beresp.http.Vary = "X-RestApi";
  }

```

**Comment by jensens** [comment]
> I have some doubt about the implementation, my test (using this branch and plone 5.2 with python 3):

Well, first, as you might see, this is just a draft. It is far from finished - it is in the middle of my work. 
Second, thanks for your testing of my half baked code and for your input! 

I will work next week(s) on this - so stay tuned.

**Comment by tisto** [comment]
@jensens can you give us a heads up about the status of the caching profile here? We discussed doing an alpha release maybe in the @plone/volto-team meeting.

**Comment by jensens** [comment]
Some notes about `Vary: ` header

- Browser respects it, but most browsers are only caching one version (the last one fetched)
- Varnish respects it if configured correctly
- Edge caches mostly not (like Cloudflare)

Notes about `Vary` on `Accept`

- pollutes the cache with several versions of the same site, because different browsers (and its version, also depending on add-ons installed) are sending very different `Accept` headers.
- does not work on edge caches (see above)
- in Browser almost useless, since only the last fetched version will be kept.

Ergo: Solution will be to not `Vary` on `Accept`, drop content negotiation on accept and use an explicit URL for RESTful requests vs. "classic" requests. See also https://github.com/plone/plone.rest/issues/87

**Comment by jensens** [comment]
- Feature wise it is working. 
- For some reason tests "test_profile_without_caching_proxy" are failing for restapi. Needs investigation.

**Comment by tisto** [comment]
@jensens just for the record. There is nothing in this PR that actually breaks anything, right. We are just adding this new profile, correct? So we could use that in Plone 5.2 core (if we would want to, I am not saying we should).

In any case, I'd go with an alpha phase for this in case we want to change things in the near future without too much hassle.

@plone/release-team do we have an official policy for Plone 6 (only) releases yet? I guess if we would go with semantic versioning and just upgrade to a minor version, we risk getting this into the 5.x branch (and we won't be able to do another feature release in 5.x).

My gut feeling would be that we should do a major release (with an alpha phase) that we use in Plone 6 only (but the package can be used in Plone 5.2 as well).

Ideally, the package should have its own tests for this use case but I am not able to put any effort into this right now.

**Comment by jensens** [comment]
> There is nothing in this PR that actually breaks anything, right. We are just adding this new profile, correct? So we could use that in Plone 5.2 core (if we would want to, I am not saying we should).

Well, the master drops Python 2 support and its based on it. Also, the caching bahaviour changes drastically for all restapi related. Thus I consider this breaking, too.

But nothing stops this working in a Plone 5.2 with Python 3. I think the classifiers are reflecting this.
Thus said, there is no CI test setup for this. Probably a GH action testing all the needed combinations would be satisfying.

**Comment by jensens** [comment]
Also, the 5.2 coredev buildout already uses the 2.x branch.

**Comment by tisto** [comment]
> Well, the master drops Python 2 support and its based on it. Also, the caching bahaviour changes drastically for all restapi related. Thus I consider this breaking, too.

Ok. Then a new major release is definitely the way to go.

**Review by tisto — APPROVED** [review]

**Comment by tisto** [comment]
Merged and released with 3.0.0a1.

**Pull request merged**